javascript/ql/lib/semmle/javascript/security/dataflow/HardcodedDataInterpretedAsCodeQuery.qll

```diff
@@ -15,7 +15,37 @@ import HardcodedDataInterpretedAsCodeCustomizations::HardcodedDataInterpretedAsC
  * A taint-tracking configuration for reasoning about hard-coded data
  * being interpreted as code
  */
-class Configuration extends TaintTracking::Configuration {
+module HardcodedDataInterpretedAsCodeConfig implements DataFlow::StateConfigSig {
+  class FlowState = DataFlow::FlowLabel;
+
+  predicate isSource(DataFlow::Node source, DataFlow::FlowLabel lbl) {
+    source.(Source).getLabel() = lbl
+  }
+
+  predicate isSink(DataFlow::Node nd, DataFlow::FlowLabel lbl) { nd.(Sink).getLabel() = lbl }
+
+  predicate isBarrier(DataFlow::Node node) { node instanceof Sanitizer }
+
+  predicate isAdditionalFlowStep(
+    DataFlow::Node node1, DataFlow::FlowLabel state1, DataFlow::Node node2,
+    DataFlow::FlowLabel state2
+  ) {
+    TaintTracking::defaultTaintStep(node1, node2) and
+    state1.isDataOrTaint() and
+    state2.isTaint()
+  }
```

Comment:
This query is a rare example of a taint-tracking configuration that uses the `data` label for the source nodes, and `taint` for the sink nodes. So it only generates an alert if at least one taint-step was used on the path.

In dataflow2, most queries don't use the `data` label at all. This query needs the step because the `data` label needs to work.